Set/ViewController.swift

```diff
@@ -8,12 +8,451 @@
+        }
+        return viewCard
+    }
+    
+    // Creates the mapper between setCard objects and their matching setCardViews
+    private func createCardsToCardsViewsMapper() {
+        
+        // When a SetGame instance is initialized, it has 69 cards in the deck and 12 cards in openCards:
+        for card in game.deck {
+            cardsToViewsMapper[card] = getCardView(of: card)
+        }
+        
+        for card in game.openCards {
+            cardsToViewsMapper[card] = getCardView(of: card)
+        }
+    }
+    
+    // Makes sure that the cardView of every open card in the game is a subview of openCardsCanvas
+    private func updateOpenCardsViews() {
+        
+        // remove from the canvas views of cards that were already matched (and aren't in openCards anymore)
+        for view in openCardsCanvas.subviews {
+            guard let index = cardsToViewsMapper.values.firstIndex(of: view) else {
+                print("Encountered an error.")
+                return
+            }
+            let card = cardsToViewsMapper.keys[index]
+            if !game.openCards.contains(card) {
+                view.removeFromSuperview()
+            }
+        }
+        
+        // adds views of cards that were recently popped from the deck to openCards
+        for card in game.openCards {
+            guard let currCardView = cardsToViewsMapper[card] else{
+                print("Couldn't find the view (cardView) for one of the open cards in the game")
+                return
+            }
+            if !openCardsCanvas.subviews.contains(currCardView) {
+                openCardsCanvas.addSubview(currCardView)
+            }
+        }
+    }
+    
+    // Initiates a turn for the player of the given player identifier.
+    private func playerButtonPressed(playerIdentifier: Int) {
+        // if this is already one of the players' turns, ignore
+        if isPlayerOneTurn || isPlayerTwoTurn {return}
+        
+        // "start a turn" for player #\(playerNumber)
+        switch playerIdentifier {
+        case 1:
+            isPlayerOneTurn = true
+            playerOneTimer = Timer.scheduledTimer(timeInterval: timeForPlayersToFindSet, target: self, selector: #selector(TimeForTurnElapsed), userInfo: nil, repeats: false)
+        case 2: // if playernumber == 2
+            isPlayerTwoTurn = true
+            playerTwoTimer = Timer.scheduledTimer(timeInterval: timeForPlayersToFindSet, target: self, selector: #selector(TimeForTurnElapsed), userInfo: nil, repeats: false)
+        default:
+            print("Provided an unsupported player identifier")
+        }
+    }
+    
+    // Make all necessary adjustments and setups in the model to display 3 new SetCardViews on the screen.
+    // (and then update the view from the model)
+    private func preformThreeCardsDealing() {
+        if isPlayerOneTurn || isPlayerTwoTurn {
+            print("Try to deal again when the turn time is over ")
```

Comment:
this should be visible to the user in an "Alert" UI form - please remember that the user doesn't see the console and when they tap a button they expect something to happen.. if they acted in unexpected way - we should let them know about it...

Reply:
👍🏻👍🏻